src/test/java/com/example/gradecalculator/service/SubjectServiceTest.java

```diff
@@ -0,0 +1,301 @@
+        when(userRepository.findById(anyLong())).thenReturn(Optional.of(user));
+        when(subjectRepository.findById(anyLong())).thenReturn(Optional.empty());
+
+        // Act & Assert
+        assertThrows(IllegalArgumentException.class, () -> subjectService.saveSubjectYearSelectedUser(1L, 1L, 1L));
+    }
+
+    @Test
+    public void testSaveSubjectYearSelectedUser_YearNotFound() {
+        // Arrange
+        when(userRepository.findById(anyLong())).thenReturn(Optional.of(user));
+        when(subjectRepository.findById(anyLong())).thenReturn(Optional.of(subject));
+        when(schoolYearRepository.findById(anyLong())).thenReturn(Optional.empty());
+
+        // Act & Assert
+        assertThrows(IllegalArgumentException.class, () -> subjectService.saveSubjectYearSelectedUser(1L, 1L, 1L));
+    }
+
+    @Test
+    public void testSaveSubjectYearSelectedUser_UserNotFound() {
+        // Arrange
+        when(userRepository.findById(anyLong())).thenReturn(Optional.empty());
+
+        // Act & Assert
+        assertThrows(IllegalArgumentException.class, () -> subjectService.saveSubjectYearSelectedUser(1L, 1L, 1L));
+    }
+
+    @Test
+    public void testSaveSubjects_WithErrors() {
+        // Arrange
+        when(subjectRepository.findById(anyLong())).thenReturn(Optional.of(subject));
+        when(schoolYearRepository.findById(anyLong())).thenReturn(Optional.of(schoolYear));
+        when(userRepository.findById(anyLong())).thenReturn(Optional.of(user));
+        when(userSubjectRepository.findByUserId(anyLong())).thenReturn(Collections.singletonList(userSubject));
+
+        String[] selectedValues = {"1-1"};
+
+        // Act
+        List<String> errors = subjectService.saveSubjects(selectedValues, 1L);
+
+        // Assert
+        assertThat(errors).isNotEmpty();
+        assertThat(errors).contains("Subject already selected for the given year: ITT2 - null");
+    }
+
+    @Test
+    public void testDeleteSubject() {
+        // Arrange
+        when(userSubjectRepository.findById(anyLong())).thenReturn(Optional.of(userSubject));
+
+        // Act
+        boolean result = subjectService.deleteSubject(1L, "1");
+
+        // Assert
+        assertTrue(result);
+        verify(userSubjectRepository, times(1)).delete(any(UserSubject.class));
+    }
+
+    @Test
+    public void testDeleteSubject_SubjectNotFound() {
+        // Arrange
+        when(userSubjectRepository.findById(anyLong())).thenReturn(Optional.empty());
+
+        // Act
+        boolean result = subjectService.deleteSubject(1L, "1");
+
+        // Assert
+        assertThat(result).isFalse();
+        verify(userSubjectRepository, never()).delete(any(UserSubject.class));
+    }
+
+    @Test
+    public void testDeleteSubject_DeleteFails() {
+        // Arrange
+        when(userSubjectRepository.findById(anyLong())).thenReturn(Optional.of(userSubject));
+//        doThrow(new RuntimeException()).when(userSubjectRepository).delete(any(UserSubject.class));
```

Comment:
?

Reply:
should not be part of this test and I forgot to delete it